service/history/hsm/tree.go

```diff
@@ -700,10 +727,85 @@ func (n *Node) root() *Node {
 	return root
 }
 
+// Compact filters the operation log to remove any transitions for deleted nodes while preserving deletion operations.
+// The returned log maintains deletion tombstones, but excludes transitions that occurred on deleted nodes or their
+// descendants.
+func (ol OperationLog) Compact() OperationLog {
+	if len(ol) == 0 {
+		return ol
+	}
+
+	root := newOpNode(Key{})
+	for _, op := range ol {
+		root.insert(op.Path(), op)
+	}
+
+	return root.collect()
+}
+
 func isPathPrefix(prefix, path []Key) bool {
 	if len(prefix) > len(path) {
 		return false
 	}
 
 	return slices.Equal(prefix, path[:len(prefix)])
 }
+
+// opNode represents a node in the operation tree. Each node maintains its operations and tracks whether it has been
+// deleted.
+type opNode struct {
+	key        Key
+	children   map[Key]*opNode
+	operations []Operation
+	isDeleted  bool
+}
+
+// newOpNode creates a new operation tree node with the given key.
+func newOpNode(key Key) *opNode {
+	return &opNode{
+		key:        key,
+		children:   make(map[Key]*opNode),
+		operations: make([]Operation, 0),
+	}
+}
+
+// insert adds an operation at the specified path in the tree.
+func (n *opNode) insert(path []Key, op Operation) {
+	if len(path) == 0 {
+		if deleteOp, ok := op.(DeleteOperation); ok {
+			n.isDeleted = true
+			for k, child := range n.children {
+				child.insert(nil, DeleteOperation{path: append(slices.Clone(deleteOp.path), k)})
+			}
+		}
+		n.operations = append(n.operations, op)
+		return
+	}
+
+	key := path[0]
+	child, exists := n.children[key]
+	if !exists {
+		child = newOpNode(key)
+		n.children[key] = child
+	}
+	child.insert(path[1:], op)
+}
+
+// collect returns an OperationLog containing all valid operations in the tree.
+func (n *opNode) collect() OperationLog {
+	var result OperationLog
+	if n.isDeleted {
+		for _, op := range n.operations {
+			if _, ok := op.(DeleteOperation); ok {
+				result = append(result, op)
+			}
+		}
+	} else {
+		result = append(result, n.operations...)
+	}
+
+	for _, child := range n.children {
```

Comment:
You should avoid adding operations for children if the node is already deleted. That's the point of compaction.

Reply:
We actually do avoid adding non-delete operations for children if the node in question is already deleted (proven by test cases). Notice that we check for "isDeleted" and, if true, we only add the DeleteOperation.

That being said, we can make this a bit cleaner. Updated!

Breakdown for clarity:

1. Tree Construction (Compact function):
  * Creates empty tree with root node
  * Take each op from log (in order) and insert into tree at proper path location

2. Operation Insertion (insert method):
  * If we're at the target path:
    * If deleted
      * mark node deleted
      * replace all ops with singular delete op
      * create cascade delete ops for all descendants
    * If not deleted
      * add op to node's op list
  * If we're not at target path:
    * Create/get next child node in path
    * Recursively inserts op with remaining path

3. Collection (collect):
* Depth-first traversal 
* Collects ops from each node in order (parent then children)
**Note: deleted nodes only contain delete ops since insertion replaced other ops**